lib/dry/schema/extensions/json_schema/schema_compiler.rb

```diff
@@ -0,0 +1,182 @@
+        attr_reader :keys, :required
+
+        # @api private
+        def initialize(root: false)
+          @keys = EMPTY_HASH.dup
+          @required = Set.new
+          @root = root
+        end
+
+        # @api private
+        def to_h
+          result = {}
+          result[:$schema] = "http://json-schema.org/draft-06/schema#" if root?
+          result.merge!(type: "object", properties: keys, required: required.to_a)
+          result
+        end
+
+        # @api private
+        def call(ast)
+          visit(ast)
+        end
+
+        # @api private
+        def visit(node, opts = EMPTY_HASH)
+          meth, rest = node
+          public_send(:"visit_#{meth}", rest, opts)
+        end
+
+        # @api private
+        def visit_set(node, opts = EMPTY_HASH)
+          target = (key = opts[:key]) ? self.class.new : self
+
+          node.map { |child| target.visit(child, opts) }
+
+          return unless key
+
+          target_info = opts[:member] ? {items: target.to_h} : target.to_h
+          type = opts[:member] ? "array" : "object"
+
+          keys.update(key => {**keys[key], type: type, **target_info})
+        end
+
+        # @api private
+        def visit_and(node, opts = EMPTY_HASH)
+          left, right = node
+
+          visit(left, opts)
+          visit(right, opts)
+        end
+
+        # @api private
+        def visit_implication(node, opts = EMPTY_HASH)
+          node.each do |el|
+            visit(el, **opts, required: false)
+          end
+        end
+
+        # @api private
+        def visit_each(node, opts = EMPTY_HASH)
+          visit(node, opts.merge(member: true))
+        end
+
+        # @api private
+        def visit_key(node, opts = EMPTY_HASH)
+          name, rest = node
+
+          if opts.fetch(:required, :true)
+            required << name.to_s
+          else
+            opts.delete(:required)
+          end
+
+          visit(rest, opts.merge(key: name))
+        end
+
+        # @api private
+        def visit_not(node, opts = EMPTY_HASH)
+          _name, rest = node
+
+          visit_predicate(rest, opts)
+        end
+
+        # @api private
+        def visit_predicate(node, opts = EMPTY_HASH)
+          name, rest = node
+          key = opts[:key]
+
+          if name.equal?(:key?)
+            prop_name = rest[0][1]
+            keys[prop_name] = {}
+          else
+            type_opts = fetch_type_opts_for_predicate(name, rest)
+            target = keys[key]
+
+            if target[:type]&.include?("array")
+              target[:items] ||= {}
+              merge_opts!(target[:items], type_opts)
+            else
+              merge_opts!(target, type_opts)
+            end
+          end
+        end
+
+        # @api private
+        def fetch_type_opts_for_predicate(name, rest)
+          type_opts = PREDICATE_TO_TYPE.fetch(name) { EMPTY_HASH }.dup
+          type_opts.transform_values! { |v| v.respond_to?(:call) ? v.call(rest[0][1]) : v }
```

Comment:
Why dup and mutate later when we can
```
PREDICATE_TO_TYPE.fetch(name, EMPTY_HASH }
  .transform_values { |v| v.respond_to?(:call) ? v.call(rest[0][1]) : v }
```
?

Reply:
it’s a style preference thing, multi line chaining feels weird to me sometimes, but happy to change if you want

Reply:
Let's wait for some @solnic (or whoever more involved in dry-schema review), we're both non-project members with our maybe-wrong opinions here after all.

Reply:
> The issue with conversion is that schemas are often used to validate incoming data and if we're using, for example, dry-schema as a source of truth and generating json-schema for some nodejs project to validate user input, even minor difference could be a security issue. Personally I wouldn't use this PR in my projects as is.
> So in this particular case IMHO we should be as strict as possible by default even if it would make feature less useful.
> I suggest to make it strict from the start and add options to _explicitly_ make it more relaxed if there would be real issues (with practical examples from users, so we could make informed decisions).

That makes sense to me. How does this sound for an interface

```ruby
my_schema.json_schema(loose: true)
```

Reply:
Something like that, yeah. But not sure about the granularity (single flag? Whitelist all constructors? Whitelist particular child schema?). 
That's why 
>with practical examples from users, so we could make informed decisions

Reply:
[`b1816a2` (#369)](https://github.com/dry-rb/dry-schema/pull/369/commits/b1816a22dbca558ca6936129bdd33ea7d667bab2)